smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2021 Adobe Systems Incorporated
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.adobe.cq.cloud.testing.it.smoke;
+
+import java.util.concurrent.TimeUnit;
+
+import com.adobe.cq.cloud.testing.it.smoke.rules.ContentPublishRule;
+import com.adobe.cq.testing.junit.rules.CQAuthorPublishClassRule;
+import com.adobe.cq.testing.junit.rules.CQRule;
+import com.adobe.cq.testing.junit.rules.Page;
+import org.apache.sling.testing.clients.util.poller.Polling;
+import org.junit.AssumptionViolatedException;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+
+/**
+ * End to end publication test.
+ * 
+ * A test page is published from author and publication is validated through the AEM
+ * publish ingress (CDN -&gt; Dispatcher -&gt; AEM publish tier).
+ */
+public class PublishEndToEndIT {
+    protected static final long TIMEOUT = TimeUnit.MINUTES.toMillis(5);
+    
+    @ClassRule
+    public static CQAuthorPublishClassRule cqBaseClassRule = new CQAuthorPublishClassRule();
+
+    @Rule
+    public CQRule cqBaseRule = new CQRule(cqBaseClassRule.authorRule, cqBaseClassRule.publishRule);
+
+    @Rule
+    public Page root = new Page(cqBaseClassRule.authorRule);
+
+    @Rule
+    public ContentPublishRule
+        contentPublishRule = new ContentPublishRule(root, cqBaseClassRule.authorRule, cqBaseClassRule.publishRule);
+
+	/**
+     * Activates a page as admin, then deactivates it.
+     * Verifies:
+     * <ul>
+     *      <li>That the replication queue is empty</li>
+     *      <li>After activation, that the page is accessible via the publish ingress</li>
+     *      <li>After deactivation, that the page is no longer accessible via the publish ingress and that the queue is empty</li>
+     * </ul>
+     *
+     * @throws Exception if an error occurred
+     */
+    @Test
+    public void testActivateAndDeactivate() throws Exception {
+
+        // This is a workaround for correctly detecting assumption violations.
+        // Any AssumptionViolatedException throw by the Callable will be
+        // swallowed by the Poller and wrapped in a TimeoutException. As a
+        // consequence, the test will be marked as failed instead of skipped.
+
+        try {
+            contentPublishRule.activateAndDeactivate(true);
+        } catch (AssumptionViolatedException e) {
```

Comment:
@amit-jain , I would do it the other way around and change the rule to throw an exception, please.

Reply:
@dulvac You want the meat of the method to move to the rule? What is the benefit you see for that? This seems to be some ad-hoc implementation to tide over a lacuna in the Poller right?

Reply:
@amit-jain , No, I'm talking about catching `AssumptionViolatedException` here from the `contentPublishRule`. Who throws the `AssumptionViolatedException` and why can't it throw an exception instead?

Reply:
From what I understand, the implementation wants to throw the `AssumptionViolatedException` to skip tests rather than fail in cases where the `AssumptionViolatedException` is thrown. The documentation over it also says that

> This is a workaround for correctly detecting assumption violations.

 The exception is thrown by the checkPage() method
`
       if (null != res && res.getStatusLine().getStatusCode() == SC_UNAUTHORIZED) {
            throw new AssumptionViolatedException("Publish requires auth for (SAML?). Skipping...");
        }
`

If we really don't care about this special handling then we can easily remove this and directly invoke the Poller.poll() warpping the method. wdyt?

Reply:
@amit-jain , we're conflating some use-cases here. We need the `AssumptionViolatedException` in checkPage to be there **and to propagate all the way through** (i.e. the test needs to be skipped in that case)

I was under the incorrect impression that the rule throws an `AssumptionViolationException` if waitPublishQueueEmptyOfPath timed out. Which I see it's not the case and it should remain like that. However, we need to not catch this `AssumptionViolationException` thrown by `checkPage` in the test.

So to recap:
* if `checkPage` throws `AssumptionViolationException`, let it go through and **skip** the test
* if `waitPublishQueueEmptyOfPath` fails, fail the test with the clear message (which you've done iiuc)

Reply:
@dulvac 
> However, we need to NOT catch this AssumptionViolationException thrown by checkPage in the test.

Yes that is also already the case. What might be confusing is the catch and rethrow. 
That as explained in the comments is to get over the limitation while wrapping the call in `Poller.poll()`. When `Poller.poll()` is callled it will after requisite retries throw only `TimeoutException` and not the underlying `AssumptionViolationException`. 

To get over that what the code is doing is making the 

- 1st call made as it is and rethrowing if its `AssumptionViolationException` to make it skip the test
- Retrying by wrapping the call in `Poller.poll()` when some other exception encountered. The call will eventually throw  a `TimeoutException` and fail if that other exception condition has not cleared.

Reply:
@amit-jain , you're right, it seems to do what it should. Indeed, I did not see the fact that **the same** exception is thrown in case of `AssumptionViolationException`. Sorry about that!